### ISSUE TITLE
Gradients (Jacobian) in inference

### ISSUE THREAD

**Issue by FelixBiermann** [pytorch/pytorch#110866]
### 🚀 The feature, motivation and pitch

Neural Nets are often applied in physical or related fields. Here, one usually not only need an output, put also its derivative. With Pytorch one can use Autograd, but I have not found a way to export a Pytorch model to onnx and then get the results from a backwards call. So if there is no way to export any Pytorch gradient calculations, is this planned for the feature? It wouldn't only be a great help for me, I am sure many others would appreciate this functionality!

(If someone found a way to achieve it, please let me know!)

**Comment by justinchuby** [comment]
Curious if the gradients are part of the forward computation (and the static graph)? Could you share an example?

**Comment by thiagocrepaldi** [comment]
`torch.onnx.export` cannot export the backward call. Using ONNX Runtime's ORTModule frontend does export your pytorch model and creates a backward graph in onnx for you (it is not the backward created by autograd). Check out https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py how to use it

One day we ight add this capability to the new onnx exporter ` torch.onnx.dynamo_export`, but this is not in our roadmap just yet.

**Comment by FelixBiermann** [comment]
> Curious if the gradients are part of the forward computation (and the static graph)? Could you share an example?

There are multiple. Every force/energy predictions or, for Taylor expansions e.g. in steady state simulations. Needing the derivatives is something probably every scientific computational software would need.

**Comment by FelixBiermann** [comment]
> `torch.onnx.export` cannot export the backward call. Using ONNX Runtime's ORTModule frontend does export your pytorch model and creates a backward graph in onnx for you (it is not the backward created by autograd). Check out https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py how to use it
> 
> One day we ight add this capability to the new onnx exporter ` torch.onnx.dynamo_export`, but this is not in our roadmap just yet.

Thank you for the link. I did look into it, but will need a bit more time to understand and ofc benchmark it. One question, how comparable is the performance between the two modes? I chose onnx because it is incredibly fast.

**Comment by thiagocrepaldi** [comment]
> > `torch.onnx.export` cannot export the backward call. Using ONNX Runtime's ORTModule frontend does export your pytorch model and creates a backward graph in onnx for you (it is not the backward created by autograd). Check out https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py how to use it
> > One day we ight add this capability to the new onnx exporter ` torch.onnx.dynamo_export`, but this is not in our roadmap just yet.
> 
> Thank you for the link. I did look into it, but will need a bit more time to understand and ofc benchmark it. One question, how comparable is the performance between the two modes? I chose onnx because it is incredibly fast.

`ORTModule` does use `torch.onnx.export  under the hood, so performance should be comparable. In fact, you can see `ORTModule` using `torch.onnx.export` as the engine for the forward step and ONNX Runtime's ONNX gradient builder engine for the backward step.

**Comment by justinchuby** [comment]
Reopening because this is an interesting topic that does not have a good solution yet.

**Comment by thiagocrepaldi** [comment]
> Reopening because this is an interesting topic that does not have a good solution yet.

@justinchuby are you planning on working on it?

Officially, `torch.onnx.export` does not support backprop and we won't be adding that as this API is being discontinued.

Also, PyTorch Dynamo in general, and thus torch.onnx.dynamo_export, doesn't support that. The exporter team won't be support in the short/medium term. Maybe in the future, when other more important features are stable.

torch.compile(backend="onnxrt") (aka DORT) should support that by default and op could try this out.

**Comment by justinchuby** [comment]
I am not planning to work on this soon, but I am interested to keep the issue open because it points to a very interesting usage.

**Comment by ianbeyst** [comment]
I can also chip in to say that this makes doing my work (MPC, I need both the Jacobian and Hessian) very difficult. I tried various things, but nothing seems to manage to produce a working exported ONNX model.

Any idea how soon "soon" is and if there is any possibility to add some priority to this?

**Comment by justinchuby** [comment]
How about using the strict=False option? Or torch.jit.trace?
